### ISSUE TITLE
DI can not resolve dipendencies required by AuthorizationHandler

### ISSUE THREAD

**Issue by Blackleones** [dotnet/aspnetcore#20185]
### Describe the bug
During the project startup, ASP.NET Core can not resolve AuthorizationHandler's dependencies which require other dependencies. 

### To Reproduce

[Sample repository](https://github.com/Blackleones/Authorizationhandler-DI-Problem)

Startup.cs 

```csharp
public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        public void ConfigureServices(IServiceCollection services)
        {
            services.AddDbContext<ApplicationDbContext>(options =>
                options.UseInMemoryDatabase("dbName"), ServiceLifetime.Transient);
            services.AddDefaultIdentity<IdentityUser>(options => options.SignIn.RequireConfirmedAccount = true)
                .AddEntityFrameworkStores<ApplicationDbContext>();
            services.AddRazorPages();
            services.AddServerSideBlazor();
            services.AddScoped<AuthenticationStateProvider, RevalidatingIdentityAuthenticationStateProvider<IdentityUser>>();
            services.AddScoped<WeatherForecastService>();
            
            services.AddTransient<ICurrentUser, CurrentUser>();
            services.AddAuthorization(opt => opt.AddPolicy("Test", p => p.Requirements.Add(new TestRequirement())));
            services.AddSingleton<IAuthorizationHandler, TestRequirementHandler>();
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            /* ... */
        }
    }
```

Definition/implementation of TestRequirement, TestRequirementHandler, ICurrentUser, CurrentUser

```csharp
public class TestRequirement : IAuthorizationRequirement
{
}

public class TestRequirementHandler : AuthorizationHandler<TestRequirement>
{
    private readonly ICurrentUser _currentUser;

    public TestRequirementHandler(ICurrentUser currentUser)
    {
        _currentUser = currentUser;
    }

    protected override Task HandleRequirementAsync(AuthorizationHandlerContext context, TestRequirement requirement)
    {
        context.Succeed(requirement);
        return Task.CompletedTask;
    }
}

public interface ICurrentUser
{
    Task<string> GetUserId();
}

public class CurrentUser : ICurrentUser
{
    private readonly AuthenticationStateProvider _provider;

    public CurrentUser(AuthenticationStateProvider provider)
    {
        _provider = provider;
    }

    public async Task<string> GetUserId()
    {
        var authState = await _provider.GetAuthenticationStateAsync();
        var user = authState.User;

        return user.Identity.IsAuthenticated ? user.Identity.Name : null;
    }
}

```

Error message:
```
Error while validating the service descriptor 'ServiceType: Microsoft.AspNetCore.Authorization.IAuthorizationService Lifetime: Transient ImplementationType: Microsoft.AspNetCore.Authorization.DefaultAuthorizationService': Cannot consume scoped service 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' from singleton 'Microsoft.AspNetCore.Authorization.IAuthorizationHandler'.
```

Stack trace:
```
at Microsoft.Extensions.DependencyInjection.ServiceProvider..ctor(IEnumerable`1 serviceDescriptors, ServiceProviderOptions options)
at Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions.BuildServiceProvider(IServiceCollection services, ServiceProviderOptions options)
at Microsoft.Extensions.DependencyInjection.DefaultServiceProviderFactory.CreateServiceProvider(IServiceCollection containerBuilder)
at Microsoft.Extensions.Hosting.Internal.ServiceFactoryAdapter`1.CreateServiceProvider(Object containerBuilder)
at Microsoft.Extensions.Hosting.HostBuilder.CreateServiceProvider()
at Microsoft.Extensions.Hosting.HostBuilder.Build()
at BlazorApp4.Program.Main(String[] args) in C:\Users\Leonardo\source\repos\BlazorApp4\BlazorApp4\Program.cs:line 18
```

### Further technical details
- ASP.NET Core version **3.1.3**
- Include the output of `dotnet --info`

```
.NET Core SDK (reflecting any global.json):
 Version:   3.1.201
 Commit:    b1768b4ae7

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18362
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.201\

Host (useful for support):
  Version: 3.1.3
  Commit:  4a9f85e9f8

.NET Core SDKs installed:
  2.1.802 [C:\Program Files\dotnet\sdk]
  2.2.402 [C:\Program Files\dotnet\sdk]
  3.0.100 [C:\Program Files\dotnet\sdk]
  3.1.102 [C:\Program Files\dotnet\sdk]
  3.1.200 [C:\Program Files\dotnet\sdk]
  3.1.201 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

- The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version: **Visual Studio 2019 Enterprise 16.5.1**

**Comment by Tratcher** [comment]
That error message says it all, but could use some interpretation.

Error while validating the service descriptor 'ServiceType: 
Microsoft.AspNetCore.Authorization.IAuthorizationService Lifetime:
Transient ImplementationType: 
Microsoft.AspNetCore.Authorization.DefaultAuthorizationService':
 Cannot consume scoped service 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' from singleton 'Microsoft.AspNetCore.Authorization.IAuthorizationHandler'.

In other words, a singleton service like TestRequirementHandler can't transitively depend on a scoped service like AuthenticationStateProvider.

@HaoK any suggestions for untangling the above scenario?

**Comment by Blackleones** [comment]
@Tratcher Thank you! it is my fault. I've changed the TestRequirementHandler service's scope from Singleton to Transient and now it works.
